concordium-consensus/src/Concordium/GlobalState/AccountMap/DifferenceMap.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE BangPatterns #-}
+
+-- | The 'DifferenceMap' stores accounts have been created in a non-finalized block.
+--  When a block is being finalized then the assoicated 'DifferenceMap' must be written
+--  to disk via 'Concordium.GlobalState.AccountMap.LMDB.insert'.
+module Concordium.GlobalState.AccountMap.DifferenceMap where
+
+import qualified Data.Map.Strict as Map
+import Prelude hiding (lookup)
+
+import Concordium.Types
+
+-- | A difference map that indicates newly added accounts for
+--  a block identified by a 'BlockHash' and its associated 'BlockHeight'.
+--  The difference map only contains accounds that was added since the '_dmParentMap'.
+data DifferenceMap = DifferenceMap
+    { -- | Accounts added in a block.
+      dmAccounts :: !(Map.Map AccountAddress AccountIndex),
+      -- | Parent map of non-finalized blocks.
+      --  In other words, if the parent block is finalized,
+      --  then the parent map is @Notnhing@ as the LMDB account map
+      --  should be consulted instead.
+      dmParentMap :: !(Maybe DifferenceMap)
+    }
+    deriving (Eq, Show)
+
+-- | Gather all accounts from the provided 'DifferenceMap' and its parent maps.
+--  Accounts are returned in ascending order of their 'AccountIndex'.
+flatten :: DifferenceMap -> [(AccountAddress, AccountIndex)]
+flatten dmap = go dmap []
+  where
+    go :: DifferenceMap -> [(AccountAddress, AccountIndex)] -> [(AccountAddress, AccountIndex)]
+    go DifferenceMap{dmParentMap = Nothing, ..} !accum =
+        let !listOfAccounts = Map.toList dmAccounts
+        in  listOfAccounts <> accum
+    go DifferenceMap{dmParentMap = Just parentMap, ..} !accum =
+        let !listOfAccounts = Map.toList dmAccounts
+        in  go parentMap $! listOfAccounts <> accum
+
+-- | Create a new empty 'DifferenceMap' based on the difference map of
+-- the parent.
+empty :: Maybe DifferenceMap -> DifferenceMap
+empty mParentDifferenceMap =
+    DifferenceMap
+        { dmAccounts = Map.empty,
+          dmParentMap = mParentDifferenceMap
+        }
+
+-- | Lookup an account in the difference map or any of the parent
+--  difference maps.
+--  Returns @Just AccountIndex@ if the account is present and
+--  otherwise @Nothing@.
+lookup :: AccountAddress -> DifferenceMap -> Maybe AccountIndex
+lookup addr = check
+  where
+    check DifferenceMap{..} = case Map.lookupGE k dmAccounts of
+        Nothing -> check =<< dmParentMap
+        Just (foundAccAddr, accIdx) ->
+            if checkEquivalence foundAccAddr
+                then Just accIdx
+                else Nothing
+    k = createAlias addr 0
```

Comment:
Yes I believe it's needed. Since we're doing a prefix lookup we need to be sure that we don't get accidentally look past the account. The account address we're looking up via might just be largest allowed alias for the account, but we've stored the address in canonical form (hash of the credential registration id).